notebooks/problem-sets/quantum_enigma_003.ipynb

```diff
@@ -0,0 +1,240 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Enigma 003 : The Four-Square Chessboard"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![University of Sherbrooke, Institute of Quantique](images/IQ_Logo.png)\n",
+    "\n",
+    "## Overview\n",
+    "\n",
+    "Watch the following video before attempting this problem set\n",
+    "\n",
+    "\n",
+    "<div class=\"youtube-wrapper\">\n",
+    "    <iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/UuVbtFXOEKQ\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Problem 1\n",
+    "\n",
+    "Here's the code for 4 people circuit.\n",
+    "\n",
+    "The enigma uses an addition modulo two like this one\n",
+    "\n",
+    "```code\n",
+    "   10\n",
+    "+  01\n",
+    "=  11\n",
+    "```\n",
+    "\n",
+    "Such addition has the interesting characteristic that the numbers can be interchanged to any order like this\n",
+    "\n",
+    "```code\n",
+    "   11\n",
+    "+  01\n",
+    "=  10\n",
+    "```\n",
+    "\n",
+    "Or\n",
+    "\n",
+    "```code\n",
+    "   11\n",
+    "+  10\n",
+    "=  01\n",
+    "```\n",
+    "\n",
+    "Meaning that adding any two of them gives the third one as an answer (this is true for any numbers). Playing with modulo two additions also have other interesting characteristics. In the enigma, adding the first number to the second is done by applying a `CNOT` between $q_4$ and $q_6$ (and $q_5$ and $q_7$). "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<!-- ::: q-block.exercise -->\n",
+    "\n",
```

Comment:
Here is the code to add

from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister
import matplotlib

#qubits 0 to 3 are the 4 squares
#qubits 4 and 5 is where the key is hidden
#qubits 6 and 7 is where the focus first lands
#qubits 8 and 9 is where the focus lands at the end which is the key location
problem_qc = QuantumCircuit(10)

#coin distribution on each square
for i in range(4):
    problem_qc.h(i)

problem_qc.barrier()

#hiding the key under one of the 4 squares
problem_qc.h(4)
problem_qc.h(5)

problem_qc.barrier()

#finding the parity of 1's on squares for which binary numbers finish by 1 and putting the answer on q5
problem_qc.cx(1, 6)
problem_qc.cx(3, 6)
problem_qc.barrier()

#finding the parity of 1's on squares for which binary numbers have a 1 as second to last digit and putting the answer on q6
problem_qc.cx(2, 7)
problem_qc.cx(3, 7)
problem_qc.barrier()

#adding modulo 2 the position of the key and the position of the focus
problem_qc.cx(4, 6)
problem_qc.cx(5, 7)
problem_qc.barrier()

#turning the right coin
problem_qc.ccx(7,6,3)
problem_qc.barrier()
problem_qc.x(6)
problem_qc.ccx(7,6,2)
problem_qc.x(6)
problem_qc.barrier()
problem_qc.x(7)
problem_qc.ccx(7,6,1)
problem_qc.x(7)
problem_qc.barrier()
problem_qc.x(6)
problem_qc.x(7)
problem_qc.ccx(7,6,0)
problem_qc.x(7)
problem_qc.x(6)
problem_qc.barrier()

#finding the parity of 1's on squares for which binary numbers finish by 1 and putting the answer on q8
problem_qc.cx(1, 8)
problem_qc.cx(3, 8)
problem_qc.barrier()

#finding the parity of 1's on squares for which binary numbers have a 1 as second to last digit and putting the answer on q9
problem_qc.cx(2, 9)
problem_qc.cx(3, 9)
problem_qc.barrier()


problem_qc.draw(output='mpl')